File: internal/engine/sstable_reader.go
package engine

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"sort"
)

type SSTableReader struct {
	file             *os.File
	Path             string
	index            []IndexEntry
	indexStartOffset uint32
	bloomStartOffset uint32
	bloom            *BloomFilter
	MinKey           string
	MaxKey           string
}

func NewSSTableReader(path string) (*SSTableReader, error) {
	f, err := os.OpenFile(path, os.O_RDONLY, 0644)
	if err != nil {
		return nil, err
	}

	// 1. Read Footer
	if _, err := f.Seek(-16, io.SeekEnd); err != nil {
		return nil, err
	}
	footerBuf := make([]byte, 16)
	if _, err := io.ReadFull(f, footerBuf); err != nil {
		return nil, fmt.Errorf("failed to read footer: %w", err)
	}

	bloomStartOffset := binary.LittleEndian.Uint32(footerBuf[0:4])
	indexStartOffset := binary.LittleEndian.Uint32(footerBuf[4:8])
	metaStartOffset := binary.LittleEndian.Uint32(footerBuf[8:12])
	magicNumber := binary.LittleEndian.Uint32(footerBuf[12:16])

	if magicNumber != 0xABCD1234 {
		f.Close()
		return nil, fmt.Errorf("invalid magic number: %x", magicNumber)
	}

	// 2. Load Bloom Filter
	bloomSize := indexStartOffset - bloomStartOffset
	bloomBytes := make([]byte, bloomSize)
	if _, err := f.Seek(int64(bloomStartOffset), io.SeekStart); err != nil {
		return nil, err
	}
	if _, err := io.ReadFull(f, bloomBytes); err != nil {
		return nil, fmt.Errorf("failed to read bloom filter: %w", err)
	}

	// 3. Load Sparse Index
	if _, err := f.Seek(int64(indexStartOffset), io.SeekStart); err != nil {
		return nil, err
	}
	countBuf := make([]byte, 4)
	if _, err := io.ReadFull(f, countBuf); err != nil {
		return nil, err
	}
	numEntries := binary.LittleEndian.Uint32(countBuf)

	index := make([]IndexEntry, numEntries)
	for i := 0; i < int(numEntries); i++ { // FIX 1: Corrected increment
		if _, err := io.ReadFull(f, countBuf); err != nil {
			return nil, err
		}
		keyLen := binary.LittleEndian.Uint32(countBuf)
		keyBuf := make([]byte, keyLen)
		if _, err := io.ReadFull(f, keyBuf); err != nil { // FIX 4: Error check
			return nil, err
		}
		if _, err := io.ReadFull(f, countBuf); err != nil {
			return nil, err
		}
		offset := binary.LittleEndian.Uint32(countBuf)

		index[i] = IndexEntry{Key: keyBuf, Offset: offset}
	}

	// 4. Load Metadata
	if _, err := f.Seek(int64(metaStartOffset), io.SeekStart); err != nil {
		return nil, err
	}
	lenBuf := make([]byte, 4)

	if _, err := io.ReadFull(f, lenBuf); err != nil {
		return nil, err
	}
	minLen := binary.LittleEndian.Uint32(lenBuf)
	minBuf := make([]byte, minLen)
	if _, err := io.ReadFull(f, minBuf); err != nil {
		return nil, err
	}

	if _, err := io.ReadFull(f, lenBuf); err != nil {
		return nil, err
	}
	maxLen := binary.LittleEndian.Uint32(lenBuf)
	maxBuf := make([]byte, maxLen)
	if _, err := io.ReadFull(f, maxBuf); err != nil {
		return nil, err
	}

	return &SSTableReader{
		file:             f,
		index:            index,
		Path:             path,
		indexStartOffset: indexStartOffset,
		bloomStartOffset: bloomStartOffset,
		bloom:            LoadBloomFilter(bloomBytes, 3),
		MinKey:           string(minBuf),
		MaxKey:           string(maxBuf),
	}, nil
}

func (r *SSTableReader) Get(userKey []byte, targetSeqNum uint64) ([]byte, bool, bool, error) {
	if !r.bloom.MightContain(string(userKey)) {
		return nil, false, false, nil
	}

	if len(r.index) == 0 {
		return nil, false, false, nil
	}

	// 2. Binary Search
	idx := sort.Search(len(r.index), func(i int) bool {
		return bytes.Compare(r.index[i].Key, userKey) >= 0
	})

	// FIX 2: Correct index navigation logic
	if idx > 0 && (idx == len(r.index) || bytes.Compare(r.index[idx].Key, userKey) > 0) {
		idx--
	}
	// If idx is 0, we check the first block. If idx > 0, we check the block starting at idx.

	blockStart := r.index[idx].Offset
	var blockEnd uint32
	if idx+1 < len(r.index) {
		blockEnd = r.index[idx+1].Offset
	} else {
		blockEnd = r.bloomStartOffset
	}

	bytesToRead := int(blockEnd - blockStart)
	blockBuf := make([]byte, bytesToRead)
	if _, err := r.file.ReadAt(blockBuf, int64(blockStart)); err != nil {
		return nil, false, false, err
	}

	// 3. Scanning with FIX 3: Correct Offset Advancement
	currentOffset := 0
	for currentOffset < bytesToRead {
		intKeyLen := int(binary.LittleEndian.Uint32(blockBuf[currentOffset : currentOffset+4]))
		currentOffset += 4

		internalKey := blockBuf[currentOffset : currentOffset+intKeyLen]
		currentOffset += intKeyLen

		valLen := int(binary.LittleEndian.Uint32(blockBuf[currentOffset : currentOffset+4]))
		currentOffset += 4

		// Capture the value slice before potentially skipping or returning
		valBytes := blockBuf[currentOffset : currentOffset+valLen]

		currUserKey, currSeqNum, currType := ParseInternalKey(internalKey)
		cmp := bytes.Compare(currUserKey, userKey)

		if cmp == 0 && currSeqNum <= targetSeqNum {
			if currType == TypeDelete {
				return nil, true, true, nil
			}
			return append([]byte(nil), valBytes...), false, true, nil
		}

		// ALWAYS advance the offset by valLen to keep the loop synchronized
		currentOffset += valLen

		if cmp > 0 {
			break
		}
	}

	return nil, false, false, nil
}

File: internal/engine2_0/sstable_reader.cpp
#include "sstable_reader.hpp"
#include "internal_key.hpp"
#include <algorithm>
#include <stdexcept>
#include <cstring>

static uint32_t Uint32LE(const uint8_t* buf) {
    return static_cast<uint32_t>(buf[0]) |
           (static_cast<uint32_t>(buf[1]) << 8) |
           (static_cast<uint32_t>(buf[2]) << 16) |
           (static_cast<uint32_t>(buf[3]) << 24);
}

SSTableReader::SSTableReader(const std::string& path)
    : path_(path), indexStartOffset_(0), bloomStartOffset_(0) {}

SSTableReader::~SSTableReader() {
    Close();
}

std::shared_ptr<SSTableReader> SSTableReader::Open(const std::string& path) {
    auto reader = std::shared_ptr<SSTableReader>(new SSTableReader(path));
    reader->file_.open(path, std::ios::in | std::ios::binary);
    if (!reader->file_.is_open()) {
        return nullptr;
    }

    // 1. Read Footer (16 bytes from end)
    reader->file_.seekg(-16, std::ios::end);
    uint8_t footerBuf[16];
    reader->file_.read(reinterpret_cast<char*>(footerBuf), 16);
    if (reader->file_.gcount() != 16) {
        return nullptr;
    }

    reader->bloomStartOffset_ = Uint32LE(footerBuf);
    reader->indexStartOffset_ = Uint32LE(footerBuf + 4);
    uint32_t metaStartOffset = Uint32LE(footerBuf + 8);
    uint32_t magicNumber = Uint32LE(footerBuf + 12);

    if (magicNumber != 0xABCD1234U) {
        return nullptr;
    }

    // 2. Load Bloom Filter
    uint32_t bloomSize = reader->indexStartOffset_ - reader->bloomStartOffset_;
    std::vector<uint8_t> bloomBytes(bloomSize);
    reader->file_.seekg(reader->bloomStartOffset_, std::ios::beg);
    reader->file_.read(reinterpret_cast<char*>(bloomBytes.data()), bloomSize);
    if (reader->file_.gcount() != bloomSize) {
        return nullptr;
    }
    reader->bloom_ = std::make_unique<BloomFilter>(std::move(bloomBytes), 3);

    // 3. Load Sparse Index
    reader->file_.seekg(reader->indexStartOffset_, std::ios::beg);
    uint8_t countBuf[4];
    reader->file_.read(reinterpret_cast<char*>(countBuf), 4);
    if (reader->file_.gcount() != 4) {
        return nullptr;
    }
    uint32_t numEntries = Uint32LE(countBuf);

    reader->index_.resize(numEntries);
    for (uint32_t i = 0; i < numEntries; i++) {
        reader->file_.read(reinterpret_cast<char*>(countBuf), 4);
        if (reader->file_.gcount() != 4) {
            return nullptr;
        }
        uint32_t keyLen = Uint32LE(countBuf);

        std::vector<uint8_t> keyBuf(keyLen);
        reader->file_.read(reinterpret_cast<char*>(keyBuf.data()), keyLen);
        if (reader->file_.gcount() != keyLen) {
            return nullptr;
        }

        reader->file_.read(reinterpret_cast<char*>(countBuf), 4);
        if (reader->file_.gcount() != 4) {
            return nullptr;
        }
        uint32_t offset = Uint32LE(countBuf);

        reader->index_[i] = IndexEntry{std::move(keyBuf), offset};
    }

    // 4. Load Metadata
    reader->file_.seekg(metaStartOffset, std::ios::beg);
    uint8_t lenBuf[4];
    
    reader->file_.read(reinterpret_cast<char*>(lenBuf), 4);
    if (reader->file_.gcount() != 4) {
        return nullptr;
    }
    uint32_t minLen = Uint32LE(lenBuf);
    std::vector<char> minBuf(minLen);
    reader->file_.read(minBuf.data(), minLen);
    if (reader->file_.gcount() != minLen) {
        return nullptr;
    }
    reader->minKey_ = std::string(minBuf.data(), minLen);

    reader->file_.read(reinterpret_cast<char*>(lenBuf), 4);
    if (reader->file_.gcount() != 4) {
        return nullptr;
    }
    uint32_t maxLen = Uint32LE(lenBuf);
    std::vector<char> maxBuf(maxLen);
    reader->file_.read(maxBuf.data(), maxLen);
    if (reader->file_.gcount() != maxLen) {
        return nullptr;
    }
    reader->maxKey_ = std::string(maxBuf.data(), maxLen);

    return reader;
}

bool SSTableReader::Get(const std::vector<uint8_t>& userKey, uint64_t targetSeqNum, std::vector<uint8_t>& val, bool& isDeleted) const {
    if (!bloom_->MightContain(std::string(userKey.begin(), userKey.end()))) {
        return false;
    }

    if (index_.empty()) {
        return false;
    }

    // Binary Search
    auto it = std::lower_bound(index_.begin(), index_.end(), userKey, [](const IndexEntry& entry, const std::vector<uint8_t>& key) {
        return entry.Key < key;
    });
    size_t idx = std::distance(index_.begin(), it);

    if (idx > 0 && (idx == index_.size() || index_[idx].Key > userKey)) {
        idx--;
    }

    uint32_t blockStart = index_[idx].Offset;
    uint32_t blockEnd = 0;
    if (idx + 1 < index_.size()) {
        blockEnd = index_[idx + 1].Offset;
    } else {
        blockEnd = bloomStartOffset_;
    }

    uint32_t bytesToRead = blockEnd - blockStart;
    std::vector<uint8_t> blockBuf(bytesToRead);

    {
        std::lock_guard<std::mutex> lock(file_mutex_);
        if (!file_.is_open()) return false;
        file_.seekg(blockStart, std::ios::beg);
        file_.read(reinterpret_cast<char*>(blockBuf.data()), bytesToRead);
        if (file_.gcount() != bytesToRead) {
            return false;
        }
    }

    size_t currentOffset = 0;
    while (currentOffset < bytesToRead) {
        if (currentOffset + 4 > bytesToRead) break;
        uint32_t intKeyLen = Uint32LE(blockBuf.data() + currentOffset);
        currentOffset += 4;

        if (currentOffset + intKeyLen > bytesToRead) break;
        std::vector<uint8_t> internalKey(blockBuf.begin() + currentOffset, blockBuf.begin() + currentOffset + intKeyLen);
        currentOffset += intKeyLen;

        if (currentOffset + 4 > bytesToRead) break;
        uint32_t valLen = Uint32LE(blockBuf.data() + currentOffset);
        currentOffset += 4;

        if (currentOffset + valLen > bytesToRead) break;
        std::vector<uint8_t> valBytes(blockBuf.begin() + currentOffset, blockBuf.begin() + currentOffset + valLen);
        
        std::vector<uint8_t> currUserKey;
        uint64_t currSeqNum = 0;
        uint8_t currType = 0;
        if (ParseInternalKey(internalKey, currUserKey, currSeqNum, currType)) {
            if (currUserKey == userKey && currSeqNum <= targetSeqNum) {
                if (currType == TypeDelete) {
                    isDeleted = true;
                    return true;
                }
                val = std::move(valBytes);
                isDeleted = false;
                return true;
            }
            if (currUserKey > userKey) {
                break;
            }
        }

        currentOffset += valLen;
    }

    return false;
}

void SSTableReader::Close() {
    std::lock_guard<std::mutex> lock(file_mutex_);
    if (file_.is_open()) {
        file_.close();
    }
}

std::string SSTableReader::GetPath() const {
    return path_;
}

std::string SSTableReader::GetMinKey() const {
    return minKey_;
}

std::string SSTableReader::GetMaxKey() const {
    return maxKey_;
}

uint32_t SSTableReader::GetBloomStartOffset() const {
    return bloomStartOffset_;
}
